fairscale/nn/moe/moe_layer.py

```diff
@@ -66,29 +66,28 @@ def __init__(self, gate: Module, experts: Union[Module, ModuleList], group: Opti
                 p.expert = True  # type: ignore
 
     def all_to_all_dispatch(self, dispatch_mask: Tensor, input: Tensor) -> Tensor:
-        dispatched_input = torch.einsum("gsec,gsm->egcm", dispatch_mask.float(), input)
+        dispatched_input = torch.einsum("sec,sm->ecm", dispatch_mask.float(), input)
         return _AllToAll.apply(self.group, dispatched_input)
 
     def all_to_all_combine(self, combine_weights: Tensor, input: Tensor) -> Tensor:
         expert_output = _AllToAll.apply(self.group, input)
-        return torch.einsum("gsec,egcm->gsm", combine_weights, expert_output)
+        return torch.einsum("sec,ecm->sm", combine_weights, expert_output)
 
     def forward(self, *input: Tensor, **kwargs: Any) -> Tensor:
         assert len(input) == 1, "only single input Tensor supported"
-        assert len(input[0].shape) == 4, "input Tensor must have dimensions: (g)roup, (s)equence, (t)oken, (m)odel"
-        assert input[0].shape[0] == len(self.experts), "group dimension size must be equal to number of local experts"
+        assert len(input[0].shape) == 3, "input Tensor must have dimensions: (s)equence, (t)oken, (m)odel"
+        assert input[0].shape[0] % len(self.experts) == 0, "num tokens must be order of number of local experts"
 
         # Implement Algorithm 2 from GShard paper.
         shape = input[0].shape
         # Reshape into S tokens per group.
-        reshaped_input = input[0].reshape(shape[0], -1, shape[3])
+        reshaped_input = input[0].reshape(-1, shape[2])
         self.l_aux, combine_weights, dispatching_mask = self.gate(reshaped_input)
         dispatched_input = self.all_to_all_dispatch(dispatching_mask, reshaped_input)
-        assert dispatched_input.shape[1] == len(self.experts)
-        chunks = dispatched_input.chunk(len(self.experts), dim=1)
+        chunks = dispatched_input.chunk(len(self.experts), dim=0)
         expert_outputs = []
         for chunk, expert in zip(chunks, self.experts):
             expert_outputs += [expert(chunk)]
-        expert_output = torch.cat(expert_outputs, dim=1)
+        expert_output = torch.cat(expert_outputs, dim=0)
```

Comment:
Fixed in a followup change. Since it is a large change, wanted to limit it to just dropping the g dimension. I can squash the changes if you think that will be easier to review.